### PR TITLE
Add new reduction interface versions of kernels

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR is am enhancement that adds reduction "tunings" for kernels with reductions. One for the RAJA default reduction interface and one for the new interface. This allows us to easily compare performance of the two RAJA reduction interfaces.
- Implementations of all kernels with reductions using the new interface have been added.
- SYCL variant of REDUCE_SUM kernel also added.
- Lastly it fixes a few compilation issues discovered when manually testing all back-ends (e.g., openmp target)

Resolves: https://github.com/LLNL/RAJAPerf/issues/339

### CODE REVIEW

File: src/algorithm/REDUCE_SUM-Cuda.cpp
Comment (MrBurmark) [inline src/algorithm/REDUCE_SUM-Cuda.cpp:226]: We should probably switch these to the with_reduce policy.

Comment (rhornung67) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:226]: Based on the discussion at the meeting this morning, I thought you wanted to hold off on changing RAJA policies for reduction kernels for now. I didn't change any policies that were already there.

Comment (MrBurmark) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:226]: We can do it later, I forgot about this particular wrinkle in the reduction tunings.

File: src/algorithm/REDUCE_SUM-OMP.cpp
Comment (MrBurmark) [inline src/algorithm/REDUCE_SUM-OMP.cpp:99]: Why not `else if`?

File: src/basic/REDUCE3_INT-OMP.cpp
Comment (MrBurmark) [inline src/basic/REDUCE3_INT-OMP.cpp:136]: Why cast here?

File: src/basic/REDUCE3_INT-OMPTarget.cpp
Comment (MrBurmark) [inline src/basic/REDUCE3_INT-OMPTarget.cpp:70]: No more old reducers?

File: src/algorithm/REDUCE_SUM-OMP.cpp
Comment (rhornung67) [reply src/algorithm/REDUCE_SUM-OMP.cpp:99]: Dunno. Just started with that and went with it.

File: src/basic/REDUCE3_INT-OMP.cpp
Comment (rhornung67) [reply src/basic/REDUCE3_INT-OMP.cpp:136]: It's probably not needed. I was trying to be explicit and follow the pattern Mike established in the RAJA tests for the new reduction stuff.

File: src/basic/REDUCE3_INT-OMPTarget.cpp
Comment (rhornung67) [reply src/basic/REDUCE3_INT-OMPTarget.cpp:70]: Crap. That's a cut-paste error. I will fix.  

Although I can't find a compiler that produced correct results for the old reduction interface with OpenMP Target. So we may want to get rid of it -- easy transition since no one uses it.

Comment (MrBurmark) [reply src/basic/REDUCE3_INT-OMPTarget.cpp:70]: ya, sounds reasonable to me, should we remove the ompt old reduction impl from RAJA at some point?

Comment (rhornung67) [reply src/basic/REDUCE3_INT-OMPTarget.cpp]: Code fix pushed. 

I think we can remove it at some point while we are cleaning up other things. Always easier to maintain less code.

**Review by MrBurmark — APPROVED** [review]

**Comment by rhornung67** [comment]
@MrBurmark this should be good to merge if you are OK with the if-else changes for the reduction tuning logic.  @artv3 or @mdavis36 do you have any comments or suggestions before I merge?

**Review by artv3 — APPROVED** [review]
Looks good to me!

**Review by MrBurmark — APPROVED** [review]

**Review by mdavis36 — APPROVED** [review]

**Pull request merged**
